app/javascript/src/utils/compiler.js

```diff
@@ -1,520 +1,570 @@
+    : source
+}
+
+function getExpressionTree(expression) {
+  expression = removeSurroundingParenthesis(expression)
+
+  const result = {}
+
+  if (expression.length === 0) {
+    result.value = ""
+    return result
+  }
+
+  for (let currentIndex = 0; currentIndex < expression.length; currentIndex++) {
+    const char = expression[currentIndex]
+    if (char === "(") {
+      const closingIndex = getClosingBracket(expression, "(", ")", currentIndex - 1)
+      if (closingIndex < 0) {
+        // parentheses are open-ended, like "(a == b"
+        result.invalid = true
+        break
+      }
+      currentIndex = closingIndex
+    } else {
+      let operatorSymbol = null
+      let operatorIndex = -1
+      for (const symbol of sortedComparisonOperatorsSymbols) {
+        const index = expression.indexOf(symbol, currentIndex)
+        if (index >= 0) {
+          operatorSymbol = symbol
+          operatorIndex = index
+          break
+        }
+      }
+
+      if (operatorSymbol == null) {
+        result.value = expression
+        break
+      }
+
+      const operator = comparisonOperators[operatorSymbol]
+
+      if (result.operator == null) {
+        const lefthand = expression.substring(0, operatorIndex)
+        const righthand = expression.substring(operatorIndex + operatorSymbol.length)
+
+        result.operator = operatorSymbol
+        result.arguments = []
+
+        if (["binary", "unary-right"].includes(operator.type)) {
+          if (lefthand.length > 0) {
+            result.arguments.push(getExpressionTree(lefthand))
+          } else {
+            result.invalid = true
+            break
+          }
+        }
+        if (["binary", "unary-left"].includes(operator.type)) {
+          if (righthand.length > 0) {
+            result.arguments.push(getExpressionTree(righthand))
+          } else {
+            result.invalid = true
+            break
+          }
+        }
+
+        break
+      }
+    }
+  }
+
+  return result
+}
+
+function evaluateExpressionTree(node) {
+  if (node.invalid) {
+    return null
+  } else if (node.value != null) {
+    return node.value.trim()
+  } else {
+    const evaluatedArguments = node.arguments.map((argument) => evaluateExpressionTree(argument))
+    const result = comparisonOperators[node.operator].eval(... evaluatedArguments)
+    return result
+  }
+}
+
+function evaluateConditionals(joinedItems) {
+  const ifStartRegex = /@if[\s\n]*\(/g
+  const startBracketRegex = /[\s\n]*\{/g
+  const elseStartRegex = /[\s\n]*@else[\s\n]*\{/g
+  const elifStartRegex = /[\s\n]*@elif[\s\n]*\(/g
+
+  let match
+  while ((match = ifStartRegex.exec(joinedItems)) != null) {
+    const openingConditionParenIndex = match.index + match[0].length - 1
+    const closingConditionParenIndex = getClosingBracket(joinedItems, "(", ")", openingConditionParenIndex - 1)
+    if (closingConditionParenIndex < 0) {
+      continue
+    }
+
+    console.log("Getting if...")

```

Comment:
Please be sure to remove all console logging statements added for the purposes of debugging.